### PR TITLE
use sql function for get_form

### PR DESCRIPTION
**Pull request by snopoke**
PR into https://github.com/dimagi/commcare-hq/pull/9298

@benrudolph making a start on these

I'm wondering if the sql functions need to be versioned for the purposes of being able to roll back?

### CODE REVIEW

File: corehq/form_processor/tests/test_form_dbaccessor.py [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import uuid
+
+from django.test import TestCase
+
+from casexml.apps.case.mock import CaseFactory, CaseBlock
```
Comment (dimagimon) [inline corehq/form_processor/tests/test_form_dbaccessor.py]: F401 'CaseFactory' imported but unused

File: corehq/form_processor/backends/sql/dbaccessors.py [hunk 2]
```diff
@@ -23,7 +23,7 @@ class FormAccessorSQL(AbstractFormAccessor):
     @staticmethod
     def get_form(form_id):
         try:
-            return XFormInstanceSQL.objects.get(form_id=form_id)
+            return XFormInstanceSQL.objects.raw('SELECT * from get_form_by_id(%s)', [form_id])[0]
```
Comment (snopoke) [inline corehq/form_processor/backends/sql/dbaccessors.py]: This is cool, as long as the columns names are correct Django can automatically create the python objects for you.

File: corehq/form_processor/tests/test_form_dbaccessor.py (hunk 1 above)
Comment (dimagimon) [inline corehq/form_processor/tests/test_form_dbaccessor.py]: F401 'CaseFactory' imported but unused

File: corehq/form_processor/backends/sql/dbaccessors.py (hunk 2 above)
Comment (czue) [inline corehq/form_processor/backends/sql/dbaccessors.py]: nice

**Comment by czue** [comment]
@snopoke when you say rolling back do you mean like a code/db migration rollback?

File: corehq/form_processor/utils/migration.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+
+def get_sql_from_file(file):
+    with open(os.path.join(SQL_FOLDER, file)) as f:
+        return f.read()
+
+
+def migrate_sql_function(name):
+    forward = get_sql_from_file('{}_create.sql'.format(name))
+    reverse = get_sql_from_file('{}_drop.sql'.format(name))
```
Comment (czue) [inline corehq/form_processor/utils/migration.py]: did you intend to commit these files? i don't see them

**Comment by snopoke** [comment]
yes. We'd only need it if a deploy failed for some reason and the old code
wasn't compatible with the new sql function. Pretty unlikely and only
necessary once we're actually using this for real.

On 26 November 2015 at 18:35, Cory Zue <notifications@github.com> wrote:

> @snopoke <https://github.com/snopoke> when you say rolling back do you
> mean like a code/db migration rollback?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/9299#issuecomment-159955617>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa

File: corehq/form_processor/utils/migration.py (hunk 3 above)
Comment (snopoke) [inline corehq/form_processor/utils/migration.py]: Yes, thanks, just added them. our .gitignore was ignorening them

**Comment by czue** [comment]
not sure if we care but it seems like either now, or very soon, we're only going to run on postgres, yeah? (guessing that we don't care since it seems ~impossible to do anything practical to support other DBs, but just noting)

**Comment by czue** [comment]
makes sense on migrations. i think we should definitely have that on our radar but no need to solve yet.

**Comment by benrudolph** [comment]
agree, at the moment we ignore the ability to rollback on failed deploys. i think the philosophy should be that new DB migrations should work on previous and new code cause even if there isn't a failed deploy we'd have down time after the migrate and before the code switch. rollback for sql is generally useful though

**Comment by snopoke** [comment]
@czue re other databases, if we stick to the `SQL` language then the functions will be cross DB compatible but I don't think its necessary to do that if its going to make things more difficult or prevent us from using more advanced features

**Comment by snopoke** [comment]
I guess if you introduce breaking changes to a function you can just create a new function instead and leave the old one there until the new one is fully in use. That would prevent having to version the functions but also allow roll-back.

**Comment by czue** [comment]
good idea on the function renames.

i don't think sqlite supports functions so that at least won't work even if we stick to plain SQL. I agree it doesn't matter though.

**Pull request merged**